[5.0] Cleanup and more

frederikja163 opened the pull request:
### Purpose of this PR

* Fixes some file indentations, and removes auto generation of certain files.
* Remove a few fixmes that were fixed, and fixes a few extra ones.
* Moves the scope logic into the indentedtextwriter class directly.
* Creates vector overloads for most methods.
* Move overloaders into seperate files.
* Created an offset overloader to deal with `glDrawElements*` and `glVertexAttribXPointer`

### Testing status

* Builds and compiles.

NogginBops commented on src/GeneratorV2/Process/Overloaders/VectorOverloader.cs:
```diff
@@ -0,0 +1,188 @@
+            for (int i = 0; i < parameters.Length; i++)
+            {
+                Parameter parameter = parameters[i];
+                if (parameter.Type is CSPointer {BaseType: CSType csType} &&
```
Nitpick: spaces between the `{` and the `}`
```cs
{ BaseType: CSType csType }
```

NogginBops replied:
This could also be done with:
```cs
parameter.Type is CSPointer pointer && pointer.BaseType is CSType csType && csType.TypeName == baseType
```
this might be easier to read as it doesn't require as much knowledge about C# pattern matching.
Personally I'm fine with this like it is.

NogginBops commented on src/GeneratorV2/Process/Overloaders/VectorOverloader.cs:
```diff
@@ -0,0 +1,188 @@
+                    // Save the new parameter.
+                    nameTable.Rename(parameter, parameter.Name + "_ptr");
+                    string vectorType = "Vector" + length + (typeChar == 'f' ? "" : typeChar);
+                    newParameters[i] = parameter with
+                    {
+                        Type = new CSRef(csType.Constant ? CSRef.Type.In : CSRef.Type.Ref, new CSType(vectorType, csType.Constant)),
+                        Length = null
+                    };
```
There are functions that don't take math types by `ref`, should we also overload them?
E.g.
```
void glUniform4f(GLint location,
                 GLfloat v0,
                 GLfloat v1,
                 GLfloat v2,
                 GLfloat v3);
```

frederikja163 replied:
This is not a vector though. The vector variant is called glUniform4fv. Which is being overloaded IIRC

NogginBops replied:
I mean both a "vectors" depending on how you look at it. `glUniform4fv` sends the data by pointer, while `glUniform4f` send the vector by value on the stack.

You only *really* need one version of the overload for opentk math types though, but it's something to consider.

NogginBops commented on src/GeneratorV2/Process/Overloaders/VectorOverloader.cs:
```diff
@@ -0,0 +1,188 @@
+                }
+            }
+
+            if (foundVector)
```
This can be replaced with `vectorParameters.Count > 0`

NogginBops commented on src/GeneratorV2/Process/Overloaders/VectorOverloader.cs:
```diff
@@ -0,0 +1,188 @@
+
+                // Get the new name for the overload.
+                int vectorStartIndexInName = vIndex - 1;
+                if (char.IsDigit(name[vIndex - 2]))
+                {
+                    vectorStartIndexInName -= 1;
+                }
```
Why do we need this check?

frederikja163 replied:
Because otherwise methods with a type between the number and the 'v' will fail. Like glUniform4fv

NogginBops replied:
Is there any guarantees that if the preceding character is a number that it's part of the vector "definition" and not just part of the function itself?

frederikja163 replied:
No there is not. But the whole vector overloading probably needs a bit of work tbh. But i think we need to rely on the name somewhat for vector overloading.

NogginBops commented on src/GeneratorV2/Process/Overloaders/VectorOverloader.cs:
```diff
@@ -0,0 +1,188 @@
+                _countParameters = countParameters;
+            }
+
+            private CsScope _scope;
```
Nitpick: To match the style of the other overloaders this should be named `Scope`.

NogginBops commented on src/GeneratorV2/Process/Overloaders/VectorOverloader.cs:
```diff
@@ -0,0 +1,188 @@
+                    }
+                    else if (parameter.Length is BinaryOperation binary)
+                    {
+                        // If we have a binary operation length we need to set the count parameter to 1.
```
Why??

NogginBops commented on src/GeneratorV2/Process/Overloaders/VectorOverloader.cs:
```diff
@@ -0,0 +1,188 @@
+
+                foreach (Parameter countParameter in _countParameters)
+                {
+                    writer.WriteLine($"{countParameter.Type.ToCSString()} {nameTable[countParameter]} = 1;");
```
Why can we set all of the count parameters to `1`?

NogginBops commented on src/GeneratorV2/Process/Processor.cs:
```diff
@@ -5,17 +5,36 @@
 {
     public static class Processor
     {
+        // /!\ IMPORTANT /!\:
+        // All return type overloaders need to run before any of the other overloaders.
+        // This is to ensure that correct scoping for the new return variables.
+        // FIXME: Maybe we dont want classes for these?
```
We probably want to add a line here about `OutToReturnOverloader`.

NogginBops commented on src/GeneratorV2/Process/Processor.cs:
```diff
@@ -119,7 +138,7 @@ public static OutputData ProcessSpec(Specification spec)
             var gles3Removes = GetRemoveEntries(features, GLAPI.GLES2);
 
-            var info = new ProcessedGLInformation(allFunctions, functionToEnumGroupsUsed, allEnumsPerAPI, allEnumGroupsToIsBitmask);
+            var info = new Processor.ProcessedGLInformation(allFunctions, functionToEnumGroupsUsed, allEnumsPerAPI, allEnumGroupsToIsBitmask);
```
Why Processor.ProcessedGLInformation and not ProcessedGLInformation?

NogginBops commented on src/GeneratorV2/Writing/Writer.cs:
```diff
@@ -384,39 +197,32 @@ private static void WriteNativeMethod(NativeFunction function, bool postfixName,
         private static void WriteOverloadMethod(Overload overload1, IndentedTextWriter indentedTextWriter, bool postfixNativeCall)
         {
-            // FIXME: This was(?) used to cull "overloads" that didn't get any .
+            // This is used to cull methods that didn't get any overloads.
```
Have we confirmed this is actually used? How many functions get filtered out here?

frederikja163 replied:
I have confirmed this, idk how many methods get filtered out. But everything breaks if it is commented out, and i think i saw atleast 10 methods being filtered out with this. I can readd the FIXME if you still think its more appropriate, but i would like to start removing the FIXME's where possible.

NogginBops commented on src/GeneratorV2/Process/Processor.cs:
```diff
@@ -184,8 +203,7 @@ private static List<RemoveEntry> GetRemoveEntries(List<Feature> features, GLAPI
             List<(string vendor, RequireEntry requireEntry)> requireEntries,
-            List<RemoveEntry> removeEntries,
-            ProcessedGLInformation glInformation)
+            List<RemoveEntry> removeEntries, Processor.ProcessedGLInformation glInformation)
```
Why `Processor.ProcessedGLInformation` and not `ProcessedGLInformation`?

NogginBops reviewed (changes requested):
Personally I like keeping all of the overloaders in the same file as it makes working with them a *lot* easier in my mind. An alternative to each overloader having a separate file we could have a `Overloaders.cs` that is separate from `Processor.cs` containing all of the overloaders.

Having all of the overloaders also creates a pressure to simplify the overloading code. At the moment a lot of the overloading code feels like it needs pass to simplify some of the checks and stuff. By having all of the overloaders in the same file the complexity issue is pretty apparent.

So I vote for reconsidering splitting the files, as it would make it easier for me.

Other than the file changes I think this looks pretty good.
I have some questions and nitpicks on the two new overloaders that should probably be addressed.

frederikja163 commented on src/GeneratorV2/Process/Overloaders/PointerToOffsetOverloader.cs:
```diff
@@ -0,0 +1,105 @@
+            {"glDrawRangeElementsBaseVertexEXT", "indices"},
+            {"glDrawRangeElementsBaseVertexOES", "indices"},
+            {"glDrawRangeElementsEXT", "indices"},
+            // {"glMultiDrawElements", "indices"},
+            // {"glMultiDrawElementsBaseVertex", "indices"},
+            // {"glMultiDrawElementsBaseVertexEXT", "indices"},
+            // {"glMultiDrawElementsEXT", "indices"},
+            // {"glMultiModeDrawElementsIBM", "indices"},
```
These are commented out because they use an array of offsets, which currently isnt being handled entirely correct. This should be pretty easy to fix. Just didnt have the time yesterday.

frederikja163 replied:
Added a FIXME in the code that says this too.

frederikja163 commented on src/GeneratorV2/Process/Overloaders/VectorOverloader.cs:
```diff
@@ -0,0 +1,188 @@
+            string name = overload.OverloadName;
+
+            // First the name is passed so we can check for the correct vector type.
+            int vIndex = name.IndexOf('v');
+            if (vIndex == -1)
+            {
+                newOverloads = null;
+                return false;
+            }
```
Right, it should be lastindexof.

NogginBops replied:
Seems weird to rely on it failing later tbh. Is there not any better checks that can be done for this?

frederikja163 replied:
Well we arent really relying on this check in the first place, we just need to know the index of the v to check the size of the vector is correct.

frederikja163 commented:
@NogginBops I edited some things, fixed most of your requests and removed the vector overloader, its not ready yet for a release. but lets get all the other features and fixes out ASAP

NogginBops commented on src/GeneratorV2/IndentedTextWriter.cs:
```diff
@@ -46,10 +46,17 @@ public void WriteLine(string str)
+        }
+
+        // Utility method for creating a cs scope "{}".
+        public Scope Scope()
```
I think it might be better to be explicit here and say `CsScope` as the `Cs` prefix is also used in other parts of the code.

NogginBops reviewed (changes requested):
This is good stuff!

I only have one *slight* naming thing I think we might want to change. 
Other than that this is looking golden!

frederikja163 commented:
@NogginBops if you dont have more stuff that needs changing i believe we are go for launch :D

NogginBops reviewed (approved):
Good stuff, let's go!
